crates/symbolicator-service/src/cache.rs

```diff
@@ -119,6 +120,174 @@ impl CacheStatus {
     }
 }
 
+/// A cache entry that either contains an object of type `T`
+/// or one of several error conditions.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum CacheEntry<T> {
+    /// A valid cache entry containing an instance of `T`.
+    AllGood(T),
+    /// The object was not found at the remote location.
+    NotFound,
+    /// The object could not be fetched from the remote source due to missing
+    /// permissions.
+    ///
+    /// The attached string contains the remote source's response.
+    PermissionDenied(String), // => whatever the server returned
+    /// The object could not be fetched from the remote source due to a timeout.
+    Timeout(Duration), // => should we return the duration after which we timed out?
+    /// The object could not be fetched from the remote source due to another problem,
+    /// like connection loss, DNS resolution, or a 5xx server response.
+    ///
+    /// The attached string contains the remote source's response.
+    DownloadError(String),
+    /// The object was fetched successfully, but is invalid in some way.
+    ///
+    /// For example, this could result from an unsupported object file or an error
+    /// during symcache conversion
+    Malformed(String),
+    /// An unexpected error in symbolicator itself.
+    InternalError,
+}
+
+impl<T> CacheEntry<T> {
+    const PERMISSION_DENIED_MARKER: &[u8] = b"permissiondenied";
+    const TIMEOUT_MARKER: &[u8] = b"timeout";
+    const DOWNLOAD_ERROR_MARKER: &[u8] = b"downloaderror";
+    const MALFORMED_MARKER: &[u8] = b"malformed";
+    const INTERNAL_ERROR_MARKER: &[u8] = b"internalerror";
```

Comment:
I don’t think we will ever persist these, as this is meant for errors that happen during persisting or reading from disk.